bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -228,28 +244,91 @@ boolean doRefresh() {
+        // The current implementations of refreshAccessToken() throws an IOException or return
+        // a valid token. This handling is future proofing just in case credentials returns null for
+        // some reason. This case was caught by a poorly coded unit test.
+        LOG.info("Refreshed the OAuth token");
+        return new HeaderCacheElement(new IOException("Could not load the token for credentials: "
+            + credentials));
       } else {
-        cacheElement = new HeaderCacheElement(newToken);
+        // Success!
+        return new HeaderCacheElement(newToken);
       }
-    } catch (IOException e) {
-      cacheElement = new HeaderCacheElement(e);
-    } catch (Exception e) {
-      cacheElement = new HeaderCacheElement(new IOException("Could not read headers", e));
-    } finally {
-      synchronized (isRefreshing) {
-        headerCache.set(cacheElement);
-        isRefreshing.set(false);
-        isRefreshing.notifyAll();
+    } catch (IOException exception) {
+      // An IOException occurred. Retry with backoff.
+      if (backoff == null) {
+        // initialize backoff.
+        backoff = retryOptions.createBackoff();
+      }
+
+      // Given the backoff, either sleep for a short duration, or terminate if the backoff has
+      // reached its configured timeout limit.
+      try {
+        RetryState retryState = getRetryState(backoff);
+        if (retryState == RetryState.PerformRetry) {
+          LOG.warn("Unexpected IOExcetpion during Credential refreshing.  Retrying", exception);
+          return refreshCredentialsWithRetry(backoff);
+        } else {
+          return new HeaderCacheElement(exception);
+        }
+      } catch (IOException e) {
+        LOG.warn("Got an exception while trying to run backoff.nextBackOffMillis()", e);
+        return new HeaderCacheElement(exception);
       }
+    } catch (Exception e) {
+      LOG.warn("Got an unexpected generic exception while trying to refresh credentials.", e);
+      return new HeaderCacheElement(new IOException("Could not read headers", e));
+    }
+
+  }
+
+  /**
+   * Sleep and/or determine if the backoff has timed out.
+   * @param backoff
+   * @return RetryState indicating the current state of the retry logic.
+   * @throws IOExcetion in some cases from {@link BackOff#nextBackOffMillis()}
+   */
+  protected RetryState getRetryState(BackOff backoff) throws IOException{
+    long nextBackOffMillis;
+    nextBackOffMillis = backoff.nextBackOffMillis();
+    if (nextBackOffMillis == BackOff.STOP) {
+      LOG.warn("Exhausted the number of retries for credentials after "
+          + this.retryOptions.getMaxElaspedBackoffMillis() + " milliseconds.");
+      return RetryState.RetriesExhausted;
+    }
+    try {
+      sleeper.sleep(nextBackOffMillis);
+      // Try to perform another call.
+      return RetryState.PerformRetry;
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted while trying to refresh credentials.");
```

Comment:
I'm not sure I want to add the InterruptedException.  It doesn't actually add any valuable information